### ISSUE TITLE
TransmissionLineEquations models built with NB cause IDA to fail

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10080]
Please check [ScalableTestSuite.Electrical.TransmissionLine.ScaledExperiments.TransmissionLineEquations_N_10](https://libraries.openmodelica.org/branches/newInst-newBackend/ScalableTestSuite_noopt/files/ScalableTestSuite_noopt_ScalableTestSuite.Electrical.TransmissionLine.ScaledExperiments.TransmissionLineEquations_N_10.sim). 

The model is initialized successfully, but then fails with
```
 ##IDA## -3 error occurred at time = 0
```
According to the `IDA documentation` -3 is IDA_ERR_FAIL, whose description is

> Error test failures occurred too many times during one internal time step or minimum step size was reached.

**Comment by casella** [comment]
As usual, you can debug with a model having N = 2, i.e.
```modelica
model Test
  extends  ScalableTestSuite.Electrical.TransmissionLine.ScaledExperiments.TransmissionLineEquations_N_10 (N = 2);
  annotation(experiment(StopTime = 4e-6, Interval=2e-9, Tolerance = 1e-8),
             __OpenModelica_commandLineOptions = "--newBackend",
             __OpenModelica_simulationFlags(s = "ida"),
             uses(ScalableTestSuite(version="2.0.3")));
end Test;
```

**Comment by casella** [comment]
Updated location to check the error: https://libraries.openmodelica.org/branches/heavy_tests/ScalableTestSuite_NB/files/ScalableTestSuite_NB_ScalableTestSuite.Electrical.TransmissionLine.ScaledExperiments.TransmissionLineEquations_N_10.sim

**Comment by casella** [comment]
Further information on the root cause of the failure. 

First of all, when compiling the C code one gets this kind of warnings:
```
Test_12jac.c:75:120: warning: format specifies type 'long' but the argument has type 'size_t' (aka 'unsigned long long') [-Wformat]
    throwStreamPrint(threadData, "Error while reading lead index list of sparsity pattern. Expected %d, got %ld", 6+1, count);
                                                                                                            ~~~        ^~~~~
                                                                                                            %zu
```
Not sure if they are relevant, but I guess they should be fixed anyway.

By turning on `LOG_NLS_V`, one gets this kind of diagnostics
```
NEWTON SOLVER STARTED! equation number:  9
maximum number of function evaluation:  100
nls status
Iteration: 1
newton step
variables
[ 1]                      der(vvol)  =                0   step =               -0   old =                0
```
This is quite odd. `der(vvol)` is determined by these two equations
```modelica
  vvol = der(vol[1]);
  //Vstep = vol[1] + 2 * Rf * Cf * der(vol[1]) + Rf ^ 2 * Cf ^ 2 * der(vvol);
  Vstep = vol[1] + 2 * (1 / w) * der(vol[1]) + 1 / w ^ 2 * der(vvol);
```
As `vol` and `vvol` are both state variables, these two equations are linear and can be solved explicitly:
```modelica
  der(vol[1]) := vvol;
  der(vvol) := w^2*(Vstep - vol[1] - 2 * (1 / w) * der(vol[1]));
```
Not sure why a nonlinear system is generated.

Furthermore, I tried to inspect that nonlinear system using the debugger, but this causes OMEdit to crash.

It seems that there is something going wrong with basic causalization. @kabdelhak, @phannebohm, please check.

**Comment by kabdelhak** [comment]
The problem is sliced states.

```
  vol[N] = cur[N] * RL;
  for i in 1:N - 1 loop
    ind * der(cur[i]) = (-res * cur[i]) - (vol[i + 1] - vol[i]) / l;
  end for;
  ```
  
These equations are defined in the model. It expects `cur` for index `1:N-1` to be a state and index `N` to be an algebraic variable. This is not possible yet and i have to find a solution for this. In the current state the compiler assumes it has to differentiate `vol[N] = cur[N] * RL;` since it assumes the whole of `cur` to be a state. This would not be problematic if i already had implemented index reduction. The current output is:
```
"Error: Internal error NBResolveSingularities.noIndexReduction failed.
(1|1) Unmatched Variables
***************************
        [DER-] (10) Real[10] $DER.cur
         slice: {9}

(1|1) Unmatched Equations
***************************
        [SCAL] (1) vol[10] = cur[10] * RL ($RES_SIM_7)
         slice: {}
"
```
Which shows what i just mentioned. Since i want to implement static index reduction in the next couple of weeks, i  will decide in the next days if i want to handle these kinds of sliced states differently or if i will just let index reduction differentiate the necessary equations. This could lead to unnecessary overhead though.

**Comment by casella** [comment]
I see. In fact, I had already fixed the issue for currents (by introducing cur_x), I can do the same for the voltages (by introducing vol_x). I guess that's the best idea at the moment. Do you agree?

**Comment by casella** [comment]
@kabdelhak, sorry, you are not looking at the right model 😅.

I have already fixed this issue on the [master branch](https://github.com/casella/ScalableTestSuite/blob/master/ScalableTestSuite/Electrical/TransmissionLine.mo), which is the one that runs in the testsuite.

These are the updated equations:
```modelica
  vvol = der(vol[1]);
  Vstep = vol[1] + 2 * (1 / w) * der(vol[1]) + 1 / w ^ 2 * der(vvol);
  vol[N] = cur[N] * RL;
  for i in 1:N - 1 loop
    cur_x[i] = cur[i];
    vol_x[i] = vol[i+1]
    cap * der(vol_x[i + 1]) = (cur[i] - cur[i + 1]) / l;
    ind * der(cur_x[i]) = (-res * cur[i]) - (vol[i + 1] - vol[i]) / l;
  end for;
```
There are N - 1 differential equations for cur_x[N - 1] in the for loop, N - 1 differential equations for vol[2:N] in the for loop, _and one differential equation for vol[1] outside the for loop_. Hence cur_x and vol are full state arrays. The vol[N] array is not fully matched by the for-loop equations, but that should not cause index problems, should it?

**Comment by phannebohm** [comment]
Arrays `vol` and `vol_x` look like they make index reduction necessary. Can you rephrase the model to only use one of those two as state?

**Comment by casella** [comment]
Sure. In fact, I can get completely rid of vol_x, that is not necessary. Will do ASAP.

**Comment by casella** [comment]
Sorry, I messed up myself. If you look at the [model on master](https://github.com/casella/ScalableTestSuite/blob/b956d494e48c8e9766142319acf0bf30c5684722/ScalableTestSuite/Electrical/TransmissionLine.mo#L102C1-L115C1) right now, these are the equations:

```modelica
      vvol = der(vol[1]);
      Vstep = vol[1] + 2 * (1 / w) * der(vol[1]) + 1 / w ^ 2 * der(vvol);
      vol[N] = cur[N] * RL;
      for i in 1:N - 1 loop
        cur[i] = cur_x[i];
        cap * der(vol[i + 1]) = (cur[i] - cur[i + 1]) / l;
        ind * der(cur_x[i]) = (-res * cur[i]) - (vol[i + 1] - vol[i]) / l;
      end for;
    initial equation
      vol = zeros(N);
      cur_x = zeros(N-1);
      vvol = 0;
```
there is no `vol_x` there, only `vol`.

**Comment by casella** [comment]
@kabdelhak any update after my last comment?

Thanks!

**Comment by phannebohm** [comment]
It's not an easy fix. We made some improvements on `TransmissionLineModelica` but we're still not there. Need to fix one thing after another until this works.

**Comment by casella** [comment]
@phannebohm, @kabdelhak any update on this?

**Comment by casella** [comment]
@kabdelhak ping!
